sled-agent/src/instance.rs

```diff
@@ -867,64 +938,188 @@ impl InstanceRunner {
+
+        let request = if let Some(params) = migrate {
+            // TODO(#6073): The migration ID is sent in the VMM registration
+            // request and isn't part of the migration target params (despite
+            // being known when the migration-start request is sent). If it were
+            // sent here it would no longer be necessary to read the ID back
+            // from the saved VMM/instance state.
+            let migration_id = self
+                .state
+                .migration_in()
+                .expect("migration ensure request entails migration in")
+                .migration_id;
+
+            let mut replace_components: std::collections::HashMap<_, _> = disks
+                .into_iter()
+                .map(|disk| {
+                    (
+                        // N.B. This ID must match the one supplied when the
+                        // instance was started.
+                        disk.id.to_string(),
+                        ReplacementComponent::CrucibleStorageBackend(
+                            disk.backend,
+                        ),
+                    )
                 })
+                .collect();
+
+            replace_components.extend(nics.into_iter().map(|nic| {
+                (
+                    // N.B. This ID must match the one supplied when the
+                    // instance was started.
+                    nic.id.to_string(),
+                    ReplacementComponent::VirtioNetworkBackend(nic.backend),
+                )
+            }));
+
+            propolis_client::types::InstanceEnsureRequest {
+                properties: self.properties.clone(),
+                init: InstanceInitializationMethod::MigrationTarget {
+                    migration_id,
+                    replace_components,
+                    src_addr: params.src_propolis_addr.to_string(),
+                },
+            }
+        } else {
+            fn id_to_device_name(id: &Uuid) -> String {
+                format!("{id}:device")
+            }
+
+            let mut spec = propolis_client::types::InstanceSpecV0 {
+                board: Board {
+                    chipset: Chipset::default(),
+                    cpus: self.vcpus,
+                    memory_mb: self.memory_mib,
+                    cpuid: None,
+                },
+                components: Default::default(),
+            };
+
+            for (name, num) in [
+                ("com1", SerialPortNumber::Com1),
+                ("com2", SerialPortNumber::Com2),
+                ("com3", SerialPortNumber::Com3),
+                ("com4", SerialPortNumber::Com4),
+            ] {
+                spec.components.insert(
+                    name.to_string(),
+                    ComponentV0::SerialPort(SerialPort { num }),
+                );
+            }
+
+            for DiskComponents { id, device, backend } in disks.into_iter() {
+                assert!(spec
```

Comment:
Some of the inserts into `components` do not assert, some do - I'm wondering about the conditions being guarded against here.